dacbench/envs/layerwise_sgd.py

```diff
@@ -0,0 +1,539 @@
+            )
+        else:
+            raise NotImplementedError(
+                f"No implementation for instance version: {self.instance_mode}"
+            )
+        # self.lr_history = deque(torch.ones(5) * math.log10(self.initial_learning_rate))
+        self.optimizer_type = torch.optim.SGD
+        self.info = {}
+        self._done = False
+
+        self.model.to(self.device)
+
+        # create param groups then feed them into optimizer
+        param_groups, self.layer_types = self._create_param_groups()
+        self.optimizer: torch.optim.Optimizer = self.optimizer_type(
+            param_groups, **self.optimizer_params,
+        )
+
+        self.n_layers = len(self.optimizer.param_groups)
+        self.learning_rates = [self.initial_learning_rate] * self.n_layers
+        self.lr_histories = [deque(torch.ones(5) * math.log10(self.initial_learning_rate))] * self.n_layers
+        # Evaluate model initially
+        train_args = [
+            self.model,
+            self.loss_function,
+            self.train_loader,
+            self.batch_size,
+            1.0,
+            self.device,
+        ]
+        losses, train_accuracy = test(*train_args)
+        self.train_loss = losses.mean()
+        self.train_accuracy = train_accuracy.mean()
+
+        test_args = [
+            self.model,
+            self.loss_function,
+            self.test_loader,
+            self.batch_size,
+            1.0,
+            self.device,
+        ]
+        test_losses, test_accuracies = test(*test_args)
+        self.test_loss = test_losses.mean()
+        self.test_accuracy = test_accuracies.mean()
+
+        val_args = [
+            self.model,
+            self.loss_function,
+            self.validation_loader,
+            self.batch_size,
+            1.0,
+            self.device,
+        ]
+        validation_loss, validation_accuracy = test(*val_args)
+
+        self.validation_loss = validation_loss.mean()
+        self.validation_accuracy = validation_accuracy.mean()
+
+        self.min_validation_loss = None
+
+        if self.epoch_mode:
+            self.average_loss = 0
+
+        return self.get_states(), {}
+
+    def _create_param_groups(self) -> tuple[list[Any], list[Any]]:
+        param_groups = []
+        layer_types = []
+        for layer in self.model.children():
+            if isinstance(layer, torch.nn.Linear | torch.nn.Conv2d):
+                param_groups.append(
+                    {
+                        "params": layer.parameters(),
+                        "lr": self.initial_learning_rate,
+                    }
+                )
+                layer_types.append(type(layer).__name__)
+        return param_groups, layer_types
+
+    def get_default_reward(self) -> torch.tensor:
+        """The default reward function.
+
+        Args:
+            _ (_type_): Empty parameter, which can be used when overriding
```

Comment:
The described parameter does not exist. Should we add it or remove the description?